core/iwasm/compilation/aot_emit_gc.c

```diff
@@ -104,4 +136,420 @@ aot_call_aot_create_func_obj(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
+                             LLVMValueRef gc_obj, LLVMValueRef heap_type,
+                             LLVMValueRef *castable)
+{
+    LLVMValueRef param_values[2], func, value, res;
+    LLVMTypeRef param_types[2], ret_type, func_type, func_ptr_type;
+
+    param_types[0] = GC_REF_TYPE;
+    param_types[1] = I32_TYPE;
+    ret_type = INT8_TYPE;
+
+    GET_AOT_FUNCTION(wasm_obj_is_type_of, 2);
+
+    /* Call function wasm_obj_is_type_of() */
+    param_values[0] = gc_obj;
+    param_values[1] = heap_type;
+    if (!(res = LLVMBuildCall2(comp_ctx->builder, func_type, func, param_values,
+                               2, "call"))) {
+        aot_set_last_error("llvm build call failed.");
+        goto fail;
+    }
+
+    *castable = res;
+
+    return true;
+fail:
+    return false;
+}
+
+bool
+aot_compile_op_ref_as_non_null(AOTCompContext *comp_ctx,
+                               AOTFuncContext *func_ctx)
+{
+    LLVMValueRef gc_obj, cmp_gc_obj;
+    LLVMBasicBlockRef check_gc_obj_succ;
+
+    GET_REF_FROM_STACK(gc_obj);
+
+    /* Check if gc object is NULL */
+    BUILD_ISNULL(gc_obj, cmp_gc_obj, "cmp_gc_obj");
+
+    ADD_BASIC_BLOCK(check_gc_obj_succ, "check_gc_obj_succ");
+    MOVE_BLOCK_AFTER_CURR(check_gc_obj_succ);
+
+    /*  Throw exception if it is NULL */
+    if (!aot_emit_exception(comp_ctx, func_ctx, EXCE_NULL_GC_REF, true,
+                            cmp_gc_obj, check_gc_obj_succ))
+        goto fail;
+
+    return true;
+fail:
+    return false;
+}
+
+bool
+aot_compile_op_i31_new(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx)
+{
+    LLVMValueRef i31_val, i31_obj;
+
+    POP_I32(i31_val);
+
+    /* Equivalent to wasm_i31_obj_new: but use I32 to represent i31_obj directly
+       (WASMI31ObjectRef)((i31_value << 1) | 1) */
+    if (!(i31_obj = LLVMBuildShl(comp_ctx->builder, i31_val, I32_ONE,
+                                 "i31_value << 1"))) {
+        aot_set_last_error("llvm build shl failed.");
+        goto fail;
+    }
+
+    if (!(i31_obj = LLVMBuildOr(comp_ctx->builder, i31_obj, I32_ONE,
+                                "((i31_value << 1) | 1)"))) {
+        aot_set_last_error("llvm build or failed.");
+        goto fail;
+    }
+
+    PUSH_REF(i31_obj);
+
+    return true;
+fail:
+    return false;
+}
+
+bool
+aot_compile_op_i31_get(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
+                       bool sign)
+{
+    LLVMValueRef i31_obj, i31_val, cmp_i31_obj, bit_30, bit_30_is_set,
+        i31_sign_val;
+    LLVMBasicBlockRef check_i31_obj_succ;
+
+    POP_REF(i31_obj);
+
+    ADD_BASIC_BLOCK(check_i31_obj_succ, "check_i31_obj_succ");
+    MOVE_BLOCK_AFTER_CURR(check_i31_obj_succ);
+
+    /* Check if i31 object is NULL, throw exception if it is */
+    BUILD_ICMP(LLVMIntEQ, i31_obj, I32_ZERO, cmp_i31_obj, "cmp_i31_obj");
+    if (!aot_emit_exception(comp_ctx, func_ctx, EXCE_NULL_GC_REF, true,
+                            cmp_i31_obj, check_i31_obj_succ))
+        goto fail;
+
+    /* i31_obj >> 1 */
```

Comment:
Should convert i31_obj from uintptr_t to uint32 first? Or the final value may be 64-bit.